stateproof/builder.go

```diff
@@ -0,0 +1,470 @@
+	}
+
+	msg, err := GenerateStateProofMessage(l, uint64(votersHdr.Round), hdr)
+	if err != nil {
+		return builder{}, err
+	}
+
+	provenWeight, err := verify.GetProvenWeight(votersHdr, hdr)
+	if err != nil {
+		return builder{}, err
+	}
+
+	var res builder
+	res.votersHdr = votersHdr
+	res.voters = voters
+	res.message = msg
+	res.Builder, err = stateproof.MakeBuilder(msg.Hash(),
+		uint64(hdr.Round),
+		provenWeight,
+		voters.Participants,
+		voters.Tree,
+		config.Consensus[votersHdr.CurrentProtocol].StateProofStrengthTarget)
+	if err != nil {
+		return builder{}, err
+	}
+
+	return res, nil
+}
+
+func (spw *Worker) initBuilders() {
+	spw.mu.Lock()
+	defer spw.mu.Unlock()
+
+	var roundSigs map[basics.Round][]pendingSig
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		roundSigs, err = getPendingSigs(tx)
+		return
+	})
+	if err != nil {
+		spw.log.Warnf("initBuilders: getPendingSigs: %w", err)
+		return
+	}
+
+	for rnd, sigs := range roundSigs {
+		if _, ok := spw.builders[rnd]; ok {
+			spw.log.Warnf("initBuilders: round %d already present", rnd)
+			continue
+		}
+		spw.addSigsToBuilder(sigs, rnd)
+	}
+}
+
+func (spw *Worker) addSigsToBuilder(sigs []pendingSig, rnd basics.Round) {
+	builderForRound, err := spw.makeBuilderForRound(rnd)
+	if err != nil {
+		spw.log.Warnf("addSigsToBuilder: makeBuilderForRound(%d): %v", rnd, err)
+		return
+	}
+	spw.builders[rnd] = builderForRound
+
+	for _, sig := range sigs {
+		pos, ok := builderForRound.voters.AddrToPos[sig.signer]
+		if !ok {
+			spw.log.Warnf("addSigsToBuilder: cannot find %v in round %d", sig.signer, rnd)
+			continue
+		}
+
+		isPresent, err := builderForRound.Present(pos)
+		if err != nil {
+			spw.log.Warnf("addSigsToBuilder: failed to invoke builderForRound.Present on pos %d - %w ", pos, err)
+			continue
+		}
+		if isPresent {
+			spw.log.Warnf("addSigsToBuilder: cannot add %v in round %d: position %d already added", sig.signer, rnd, pos)
+			continue
+		}
+
+		if err := builderForRound.IsValid(pos, &sig.sig, false); err != nil {
+			spw.log.Warnf("addSigsToBuilder: cannot add %v in round %d: %v", sig.signer, rnd, err)
+			continue
+		}
+		if err := builderForRound.Add(pos, sig.sig); err != nil {
+			spw.log.Warnf("addSigsToBuilder: error while adding sig. inner error: %w", err)
+			continue
+		}
+	}
+}
+
+func (spw *Worker) handleSigMessage(msg network.IncomingMessage) network.OutgoingMessage {
+	var ssig sigFromAddr
+	err := protocol.Decode(msg.Data, &ssig)
+	if err != nil {
+		spw.log.Warnf("spw.handleSigMessage(): decode: %v", err)
+		return network.OutgoingMessage{Action: network.Disconnect}
+	}
+
+	fwd, err := spw.handleSig(ssig, msg.Sender)
+	if err != nil {
+		spw.log.Warnf("spw.handleSigMessage(): %v", err)
+	}
+
+	return network.OutgoingMessage{Action: fwd}
+}
+
+func (spw *Worker) handleSig(sfa sigFromAddr, sender network.Peer) (network.ForwardingPolicy, error) {
```

Comment:
this is a bit weird to return `ForwardingPolicy` from a signature validation method.
Suggest to return (forward boo, err error):
err != nil => disconnect
forward != true => ignore
else => broadcast